cerealize/cerealize.h

```diff
@@ -140,84 +140,124 @@ WithBaseType(const ENTRY& object) {
   }
+
+ protected:
+  std::ofstream fo_;
+  size_t entries_written_ = 0;
+
+ private:
+  size_t initial_stream_position_;
+
+  CerealFileAppenderBase() = delete;
+  CerealFileAppenderBase(const CerealFileAppenderBase&) = delete;
+  void operator=(const CerealFileAppenderBase&) = delete;
+  CerealFileAppenderBase(CerealFileAppenderBase&&) = delete;
+  void operator=(CerealFileAppenderBase&&) = delete;
 };
 
-// `CerealFileAppender` appends cereal-ized records to a file.
-// The format is selected by a template parameter, defaults to binary. All formats are supported.
+// `CerealBinaryFileAppender` appends cereal-ized records to a file in binary format.
 // Writes are performed using templated `operator <<(const T& entry)`.
-// Is type `T` defines a typedef of `CEREAL_BASE_TYPE`, polymorphic serialization is used.
-template <CerealFormat T_CEREAL_FORMAT>
-class GenericCerealFileAppender {
+// If type `T` defines a typedef of `CEREAL_BASE_TYPE`, polymorphic serialization is used.
+class CerealBinaryFileAppender : public CerealFileAppenderBase {
  public:
-  explicit GenericCerealFileAppender(const std::string& filename, bool append = true)
-      : fo_(filename, (append ? std::ofstream::app : std::ofstream::trunc) | std::ofstream::binary),
-        so_(CerealStreamType<T_CEREAL_FORMAT>::CreateOutputArchive(fo_)) {}
+  explicit CerealBinaryFileAppender(const std::string& filename, bool append = true)
+      : CerealFileAppenderBase(filename, append),
+        so_(cereal::BinaryOutputArchive(fo_)) {}
 
   template <typename T>
-  typename std::enable_if<sizeof(typename T::CEREAL_BASE_TYPE) != 0, GenericCerealFileAppender&>::type
+  typename std::enable_if<sizeof(typename T::CEREAL_BASE_TYPE) != 0, CerealBinaryFileAppender&>::type
   operator<<(const T& entry) {
     so_(WithBaseType<typename T::CEREAL_BASE_TYPE>(entry));
+    entries_written_++;
     return *this;
   }
 
  private:
-  GenericCerealFileAppender() = delete;
-  GenericCerealFileAppender(const GenericCerealFileAppender&) = delete;
-  void operator=(const GenericCerealFileAppender&) = delete;
-  GenericCerealFileAppender(GenericCerealFileAppender&&) = delete;
-  void operator=(GenericCerealFileAppender&&) = delete;
+  cereal::BinaryOutputArchive so_;
+};
 
-  std::ofstream fo_;
-  typename CerealStreamType<T_CEREAL_FORMAT>::Output so_;
+// `CerealJSONFileAppender` appends cereal-ized records to a file in JSON format.
+// Each entry is written as a separate line containing full JSON record.
+// Writes are performed using templated `operator <<(const T& entry)`.
+// If type `T` defines a typedef of `CEREAL_BASE_TYPE`, polymorphic serialization is used.
+class CerealJSONFileAppender : public CerealFileAppenderBase {
+ public:
+  explicit CerealJSONFileAppender(const std::string& filename, bool append = true)
+      : CerealFileAppenderBase(filename, append) {}
+
+  template <typename T>
+  typename std::enable_if<sizeof(typename T::CEREAL_BASE_TYPE) != 0, CerealJSONFileAppender&>::type
```

Comment:
Also, it would cover the case where the entry to cerealize is passed as `WithBaseType<>`, right?

Which reminds me: `WithBaseType<WithBaseType<...>...>` should not compile. Could you check real quick? I could do it too. Thanks!